libsolidity/formal/SymbolicTypes.cpp

```diff
@@ -458,6 +494,30 @@ smtutil::Expression zeroValue(frontend::TypePointer const& _type)
 	return 0;
 }
 
+bool isSigned(TypePointer const& _type)
```

Comment:
Can this be somehow moved to `Type`, at least as a static method?

Reply:
I'm not sure the semantics align. Solidity moved away from considering `address/contract` the same as numbers, but the SMTChecker still does consider that. For me it'd be no problem moving it there, but I'd like to get other opinions. @chriseth?

Reply:
I would be fine with the folloing:
Call `_type->mobileType()` here. After that, everything except InteregType and FixedPointType should return false.
Then introduce `ArithmeticType` as base class of those two and add a `isSigned()` function there.

Reply:
The semantics are slightly different though. My function also returns true for negative RationalNumberTypes. But I can also do that separately.

Reply:
Maybe I misunderstood it, but what's the point in calling `mobileType()`? It would actually only simplify one case, `RationalNumberType`. The other cases would still need to be there.

Reply:
Exactly, for all the types in that function except for `RationalNumberType`. So the function wouldn't actually change by using `mobileType`.

Reply:
If you introduce ArithmeticType as base type, and implement isSigned there it would simplify this. 
I am fine leaving as is, and implement ArithmeticType in a separate PR.

Reply:
That's what I'm not sure of. If `ArithmeticType` is implemented as base for `IntegerType` and `FixedPointType` this function could be simplified to:
```
  bool isSigned(TypePointer const& type)
  {
      auto _type = type->mobileType();
      solAssert(smt::isNumber(*_type), "");
      bool isSigned = false;
    if (auto const* aType = dynamic_cast<ArithmeticType const*>(_type))
          isSigned |= intType->isSigned();
    else if (dynamic_cast<AddressType const*>(_type))
          return false;
    else if (dynamic_cast<ContractType const*>(_type))
          return false;
    else if (dynamic_cast<EnumType const*>(_type))
          return false;
    else if (dynamic_cast<FixedBytesType const*>(_type))
          return false;
    else
          solAssert(false, "");
  
      return isSigned;
  }
```
`RationalNumberType` is handled by `mobileType()`. I still need to take care of the other types, so this function would still be there, and I'm not sure the AST changes would be worth it.